### ISSUE TITLE
Frequent connection issues

### ISSUE THREAD

**Issue by breakthestatic** [fondberg/spotcast#503]
### The problem

Since upgrading to v5, I've been seeing somewhat frequent batches of connection errors. Although anecdotal, I verified that there were no larger-scale network connectivity issues at the times that Spotcast requests were timing out. The errors are *somewhat* sporadic, but happen at least a few times a day.

As a test, I ran a second instance of Spotcast using v4 in parallel (on a separate Home Assistant install and using a different Spotify account/app). The v4 instance did not record *any* errors during the same time period that v5 was logging errors (over a roughly 36 hr period). These intitial logs were, regrettably, not captured using debug logging. However, I have another test running right now with debug enabled. Hopefully that will provide further details to help track down the issues.

Given the frequency of the issue (several times a day), I doubt that the Spotify API itself is going down. I did see some 429 errors, which indicates it might be polling too frequently. There's also some 502, 503 & 504 errors, seemingly around the token refresh logic.

## System Information

version | core-2024.12.1
-- | --
installation_type | Unsupported Third Party Container
dev | false
hassio | false
docker | true
user | homeassistant
virtualenv | true
python_version | 3.13.0
os_name | Linux
os_version | 6.1.0-28-amd64
arch | x86_64
timezone | America/Los_Angeles
config_dir | /config

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
GitHub Content | ok
GitHub Web | ok
HACS Data | ok
GitHub API Calls Remaining | 5000
Installed Version | 2.0.1
Stage | running
Available Repositories | 1490
Downloaded Repositories | 24

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | false
-- | --
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Dashboards</summary>

dashboards | 1
-- | --
resources | 19
views | 22
mode | storage

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | December 2, 2024 at 6:53 PM
-- | --
current_recorder_run | December 11, 2024 at 10:30 PM
estimated_db_size | 509.06 MiB
database_engine | sqlite
database_version | 3.45.3

</details>

<details><summary>Spotcast</summary>

Version | 5.0.0-b16
-- | --
Breakthestatic Is Default | true
Breakthestatic Public Endpoint | ok
Breakthestatic Public Token | healthy
Breakthestatic Private Endpoint | ok
Breakthestatic Private Token | healthy

</details>

<details><summary>Spotify</summary>

api_endpoint_reachable | ok
-- | --

</details>

### What type of installation are you running?

Home Assistant Container

### Diagnostics information

[home-assistant_2024-12-13T17-45-11.949Z.log](https://github.com/user-attachments/files/18129786/home-assistant_2024-12-13T17-45-11.949Z.log)

**Comment by fcusson** [comment]
@breakthestatic anything `5##` is server side issue, that's not something we can control. That could be an issue with the Spotify relay server you are connecting to that is the cause of the problem. All `5##` errors you mentioned `502 - Bad Gateway`, `503 - Service Unavailable` and `504 - Gateway Timeout` relate to issues that would be expected when a gateway server doesn't get a proper reply from its upstream.

I can confirm I have been running live Spotcast on my main server and often running concurrently my development container for long period of time without any error logs. Unless others are also running in the same issue, this look like a specific to your situation. If that's ok with you could you provide a `dig` or `nslookup` reply for `open.spotify.com`. That could help identify which gateway you are using and if any outage specific to that gateway can be found.

On the other side, the errors you are logging are mishandled message decoding when receiving `5##` error message (I was expecting `json` data but received `txt`), that is something that needs to be fixed in the code. I'll get on that during the weekend

**Comment by fcusson** [comment]
@breakthestatic 

> I was thinking that perhaps the underlying code is mis-using the "private" Spotify API in some subtle way that induces the errors. Is v4 very different in terms of the specific internal endpoints being hit and/or is the polling interval drastically different?

Did you have the sensors activated when you did the test with `V4`. In terms of api usage, its very similar, in term of the type of request being made, just maybe a bit more request because of the sensors, but I am caching the result to get the information. In total the request should amount to around 10 requests/min. In your case, you are getting the error specifically when trying to get a token using the private cookies.

**Comment by cdonovan** [comment]
I don't think any sensors were activated on v4, unless they are enabled by default. I primarily just use the service to play media and haven't looked into the other features.  In any event, both the v4 and v5 installs were essentially untouched configs just with different accounts and keys used during initial setup. Also, just for clarity, I do use both the Spotify web app and Lyrion (formerly Logitech/Squeezebox) server connected to Spotify as well. I don't think it likely would interfere but both of those are essentially running parallel sessions from the same public-facing ip too.

**Comment by fcusson** [comment]
ok so in the case of v4, that would probably be because you didn't have any sensor active, you weren't pulling any information from the API on a regular interval.

**Comment by fcusson** [comment]
Released a new build with some improvements to `5##`. For the moment this will mark the credentials as expired, but will look into ways of marking the config not ready to force retry after a delay.

**Comment by fcusson** [comment]
#504 reporting excesive expirery of tokens when `5##` errors raised. Merging the 2 tickets

**Comment by fcusson** [comment]
build 18 pushed with a filter on 500 range error. Currently this will only raise the error, failing to update sensor, but the system will not wait for a retry at the moment. This could mean high error log count during Spotify gateway outage.

**Comment by fcusson** [comment]
@breakthestatic, @rwjack

just released build 19 with a retry supervisor for Upstream Gateway error. This will not fix any `5##` range error, as these error are not possible to manage from the Home Assistant side.

Now, the integration should cleanly enter a wait and retry phases and only attempt to reconnect to the server every 30 seconds while only logging an error once at the start.

Please provide any detail if other errors or tracebacks are logged in Home Assistant so I can take care of these cases.

**Comment by fcusson** [comment]
Closing for inactivity, please open new ticket for follow-up issues
